File: composition/Scala/composition-pipeline/pipeline/src/main/scala/shared/StringExtensions.scala
package shared

import common.StringExtensions._

// This is an example of extending a built in object so that we can easily access shared business logic.
object StringExtensions {
  private val emailRegEx = """^[^@]+@[^@]+\.[^@]+$""".r

  implicit class StringLogic(val s: String) {
    def isValidEmailAddress = s == !s.isNullOrEmpty || emailRegEx.findFirstMatchIn(s).isDefined
  }

}

File: composition/C#/CompositionPipeline/Shared/StringExtensions.cs
using System.Text.RegularExpressions;

namespace CompositionPipeline.Shared
{
	public static class StringExtensions
	{
		const string emailRegEx = "^[^@]+@[^@]+\\.[^@]+$";

		public static bool IsValidEmailAddress(this string s) {
			return !string.IsNullOrWhiteSpace(s) && Regex.IsMatch(s, emailRegEx);
		}
	}
}
